Fix 'mempool min fee not met' debug output

Empact opened the pull request:
Output the value that is tested, rather than the unmodified fee value.

Prompted by looking into: #11955

Empact committed:
Fix 'mempool min fee not met' debug output

Output the value that is tested, rather than the unmodified fee value.

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/12356/commits/875d62f42a614cf03ac858a493f105f846f3afcf

promag commented:
utACK 875d62f.

It would be nice to add tests for `mempool min fee not met` and `mempool full` errors, it's the missing cases of `REJECT_INSUFFICIENTFEE`.

Empact committed:
Add test for 'mempool min fee not met' rpc error

Empact committed:
Consistently use FormatStateMessage in RPC error output

This will include the error code and debug output as well as the reason string.

See #11955 for the motivation.

Empact commented:
I added a change to include the full FormatStateMessage output in RPC errors that are currently including just the reason or reason + code. That would include the missing debug output that comes up e.g. in #11955

promag commented on test/functional/mempool_limit.py:
```diff
@@ -50,5 +50,18 @@ def run_test(self):
+        inputs = [{ "txid" : us0["txid"], "vout" : us0["vout"]}]
+        outputs = {self.nodes[0].getnewaddress() : 0.0001}
+        tx = self.nodes[0].createrawtransaction(inputs, outputs)
+        self.nodes[0].settxfee(relayfee) # specifically fund this tx with a fee < mempoolminfee, >= than minrelaytxfee
```
I believe you can set `feeRate` option in `fundrawtransaction` instead.

promag commented on test/functional/mempool_limit.py:
```diff
@@ -50,5 +50,18 @@ def run_test(self):
+        tx_hex = self.nodes[0].signrawtransaction(txF['hex'])['hex']
+        tx_id = self.nodes[0].decoderawtransaction(tx_hex)["txid"]
+        assert_raises_rpc_error(-26, "66: mempool min fee not met", self.nodes[0].sendrawtransaction, tx_hex)
+        assert(tx_id not in self.nodes[0].getrawmempool())
```
I guess there is no need to assert this, `sendrawtransaction` above failed. This `assert` makes sense when testing `sendrawtransaction` failure, so move it there if it's missing.

Empact replied:
Fair enough, removed that and generally simplified the test.

Fyi I was following the example set in: https://github.com/bitcoin/bitcoin/blob/master/test/functional/mining_prioritisetransaction.py#L124

promag reviewed (commented):
Make "Consistently use FormatStateMessage in RPC error output" commit the first (or second) so that the new test is not changed.

laanwj commented:
utACK https://github.com/bitcoin/bitcoin/pull/12356/commits/bb00c95c16f50c5dfab1aa8fbb6c873318a6acc8

promag commented:
utACK bb00c95. Thanks for the commit reorder @Empact.

MarcoFalke commented on test/functional/mempool_limit.py:
```diff
@@ -50,5 +50,15 @@ def run_test(self):
+        # specifically fund this tx with a fee < mempoolminfee, >= than minrelaytxfee
+        txF = self.nodes[0].fundrawtransaction(tx, {'feeRate': relayfee})
+        txFS = self.nodes[0].signrawtransaction(txF['hex'])
+        assert_raises_rpc_error(-26, "mempool min fee not met, 166 < 411 (code 66)", self.nodes[0].sendrawtransaction, txFS['hex'])
```
The transaction is created non-deterministically, thus you can't hardcode the exact value here.

See for example the following failure (off-by-two):

```
AssertionError: Expected substring not found:mempool min fee not met, 166 < 409 (code 66)

Empact replied:
Ok, will send up a fix shortly.
